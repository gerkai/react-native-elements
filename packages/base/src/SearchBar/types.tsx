import {
  PressableProps,
  ViewStyle,
  StyleProp,
  TextStyle,
  ActivityIndicatorProps,
} from 'react-native';
import { Theme } from '../helpers';
import { IconNode } from '../Icon';
import { InputProps } from '../Input';

export type SearchBarAndroidProps = SearchBarBaseProps & {
  platform?: 'android';
  cancelIcon?: IconNode;
  onCancel?: () => any;
};

export type SearchBarIosProps = SearchBarBaseProps & {
  platform?: 'ios';
  cancelButtonProps?: Partial<PressableProps> & {
    buttonStyle?: StyleProp<ViewStyle>;
    buttonTextStyle?: StyleProp<TextStyle>;
    color?: string;
    buttonDisabledStyle?: StyleProp<ViewStyle>;
    buttonDisabledTextStyle?: StyleProp<ViewStyle>;
  };
  cancelButtonTitle?: string;
  showCancel?: boolean;
  searchIcon?: IconNode;
  clearIcon?: IconNode;
  /**
   * @type Callback Function on cancel icon press
   */
  onCancel?: () => any;
};

export type SearchBarDefaultProps = SearchBarBaseProps & {
  platform?: 'default';
  lightTheme?: boolean;
  round?: boolean;
};

export interface SearchBarBaseProps extends InputProps {
  containerStyle?: StyleProp<ViewStyle>;
  inputContainerStyle?: StyleProp<ViewStyle>;
  clearIcon?: IconNode;
  searchIcon?: IconNode;
  inputStyle?: StyleProp<TextStyle>;
  loadingProps?: ActivityIndicatorProps;
  showLoading?: boolean;
  leftIconContainerStyle?: StyleProp<ViewStyle>;
  rightIconContainerStyle?: StyleProp<ViewStyle>;
  /**
   * @type Callback Function on clear icon press
   */
  onClear?(): void;
  theme?: Theme;
}

export type SearchBarProps =
  | SearchBarDefaultProps
  | SearchBarAndroidProps
  | SearchBarIosProps;
