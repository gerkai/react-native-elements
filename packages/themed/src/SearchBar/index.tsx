import { withTheme } from '../config';
import {
  SearchBar,
  SearchBarBaseProps,
} from '@react-native-elements/base/dist/SearchBar/SearchBar';
import { SearchBarAndroidProps } from '@react-native-elements/base/dist/SearchBar/SearchBar-android';
import { SearchBarIosProps } from '@react-native-elements/base/dist/SearchBar/SearchBar-ios';
import { SearchBarDefaultProps } from '@react-native-elements/base/dist/SearchBar/SearchBar-default';

export { SearchBar };
export type {
  SearchBarBaseProps as SearchBarProps,
  SearchBarAndroidProps,
  SearchBarDefaultProps,
  SearchBarIosProps,
};
export default withTheme<SearchBarBaseProps>(SearchBar, 'SearchBar');
