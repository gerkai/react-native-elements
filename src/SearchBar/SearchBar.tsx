import React from 'react';
import { SearchBarIOS, SearchBarIosProps } from './SearchBar-ios';
import { SearchBarAndroid, SearchBarAndroidProps } from './SearchBar-android';
import { SearchBarDefault, SearchBarDefaultProps } from './SearchBar-default';
import {
  ActivityIndicatorProps,
  StyleProp,
  TextStyle,
  ViewStyle,
  TextInput,
} from 'react-native';
import { IconNode } from '../Icon';
import { ThemeProps } from '../config';

const SEARCHBAR_COMPONENTS = {
  ios: SearchBarIOS,
  android: SearchBarAndroid,
  default: SearchBarDefault,
};

export type SearchBarBaseProps = React.ComponentPropsWithRef<
  typeof TextInput
> & {
  platform?: 'default' | 'ios' | 'android';
  containerStyle?: StyleProp<ViewStyle>;
  inputContainerStyle?: StyleProp<ViewStyle>;
  clearIcon?: IconNode;
  searchIcon?: IconNode;
  inputStyle?: StyleProp<TextStyle>;
  loadingProps?: ActivityIndicatorProps;
  showLoading?: boolean;
  leftIconContainerStyle?: StyleProp<ViewStyle>;
  rightIconContainerStyle?: StyleProp<ViewStyle>;
  onClear?(): void;
  onFocus?(): void;
  onBlur?(): void;
  onChangeText?(text: string): void;
  onCancel?(): void;
};

export type SearchBarProps =
  | SearchBarBaseProps
  | SearchBarDefaultProps
  | SearchBarAndroidProps
  | SearchBarIosProps;

export class SearchBar extends React.Component<
  SearchBarProps & Partial<ThemeProps<SearchBarProps>>
> {
  searchbar!: SearchBarIOS;
  static defaultProps = {
    platform: 'default' as const,
  };

  focus = () => {
    this.searchbar.focus();
  };

  blur = () => {
    this.searchbar.blur();
  };

  clear = () => {
    this.searchbar.clear();
  };

  cancel = () => {
    this.searchbar.cancel && this.searchbar.cancel();
  };

  render() {
    const Component: typeof React.Component =
      SEARCHBAR_COMPONENTS[this.props.platform] || SearchBarDefault;

    return (
      <Component
        ref={(ref: SearchBarIOS) => {
          this.searchbar = ref;
        }}
        {...this.props}
      />
    );
  }
}
