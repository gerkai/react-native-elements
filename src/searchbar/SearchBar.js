import PropTypes from 'prop-types';
import React, { Component } from 'react';
import IOSSearchBar from './SearchBar-ios';
import AndroidSearchBar from './SearchBar-android';
import DefaultSearchBar from './SearchBar-default';

const SEARCHBAR_COMPONENTS = {
  ios: IOSSearchBar,
  android: AndroidSearchBar,
  default: DefaultSearchBar,
};

class SearchBar extends Component {

  focus = () => {
    this.searchbar.focus();
  };

  blur = () => {
    this.searchbar.blur();
  };

  clear = () => {
    this.searchbar.clear();
  };

  cancel = () => {
    this.searchbar.cancel && this.searchbar.cancel();
  };


  render() {
    const SearchBar = SEARCHBAR_COMPONENTS[this.props.platform] || DefaultSearchBar;
    return <SearchBar ref={searchbar => this.searchbar = searchbar} {...this.props} />;
  }
}

SearchBar.propTypes = {
  platform: PropTypes.oneOf(['default', 'ios', 'android']),
};

SearchBar.defaultProps = {
  platform: 'default',
};


export default SearchBar;