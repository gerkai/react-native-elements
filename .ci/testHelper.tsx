import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeProvider, FullTheme } from '../src/config';

const theme = {
  colors: {
    black: 'black',
  },
};

export const renderWithTheme = (
  children: any,
  themeProp: Partial<FullTheme> = theme
) => {
  return render(<ThemeProvider theme={themeProp}>{children}</ThemeProvider>);
};
