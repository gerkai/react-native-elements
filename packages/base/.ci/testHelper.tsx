import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';
import { ThemeProvider, FullTheme, colors } from '../src/config';
import deepmerge from 'deepmerge';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  themeProp: Partial<FullTheme> = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(
    themeProp ? (
      <ThemeProvider theme={deepmerge({ colors }, themeProp)}>
        {children}
      </ThemeProvider>
    ) : (
      children
    ),
    options
  );
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper');
  return { wrapper, ...renderApi };
};
