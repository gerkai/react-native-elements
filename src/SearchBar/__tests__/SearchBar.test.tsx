import React from 'react';
import SearchBar from '../index';
import { renderWithTheme } from '../../../.ci/testHelper';

describe('SearchBar wrapper component', () => {
  it('should render without issues', () => {
    const component = renderWithTheme(<SearchBar />);
    expect(component).not.toBeNull();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render an iOS SearchBar', () => {
    const component = renderWithTheme(<SearchBar platform="ios" />);
    expect(component).not.toBeNull();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render an Android SearchBar', () => {
    const component = renderWithTheme(<SearchBar platform="android" />);
    expect(component).not.toBeNull();
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should apply values from theme', () => {
    const theme = {
      SearchBar: {
        placeholder: 'Enter search term',
      },
    };
    const component = renderWithTheme(<SearchBar platform="android" />, theme);
    expect(component.queryByTestId('RNE__SearchBar').props.placeholder).toBe(
      'Enter search term'
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
