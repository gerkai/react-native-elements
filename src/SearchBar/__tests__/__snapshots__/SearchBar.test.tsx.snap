// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar wrapper component should apply values from theme 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "paddingBottom": 8,
        "paddingTop": 8,
      }
    }
    testID="RNE__SearchBar-wrapper"
  >
    <View
      style={
        Object {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
      testID="RNE__Input__view-wrapper"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              replaceTheme={[Function]}
              testID="RNE__ICON__CONTAINER_ACTION"
              updateTheme={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "rgba(0, 0, 0, 0.54)",
                        "fontSize": 25,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
        <TextInput
          allowFontScaling={true}
          editable={true}
          onBlur={[Function]}
          onCancel={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Enter search term"
          placeholderTextColor="#86939e"
          platform="android"
          rejectResponderTermination={true}
          style={
            Object {
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 24,
              "marginRight": 8,
              "minHeight": 40,
            }
          }
          testID="RNE__SearchBar"
          underlineColorAndroid="transparent"
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`SearchBar wrapper component should render an Android SearchBar 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "paddingBottom": 8,
        "paddingTop": 8,
      }
    }
    testID="RNE__SearchBar-wrapper"
  >
    <View
      style={
        Object {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
      testID="RNE__Input__view-wrapper"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              replaceTheme={[Function]}
              testID="RNE__ICON__CONTAINER_ACTION"
              updateTheme={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "rgba(0, 0, 0, 0.54)",
                        "fontSize": 25,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
        <TextInput
          allowFontScaling={true}
          editable={true}
          onBlur={[Function]}
          onCancel={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholderTextColor="#86939e"
          platform="android"
          rejectResponderTermination={true}
          style={
            Object {
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 24,
              "marginRight": 8,
              "minHeight": 40,
            }
          }
          testID="RNE__SearchBar"
          underlineColorAndroid="transparent"
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;

exports[`SearchBar wrapper component should render an iOS SearchBar 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flexDirection": "row",
        "overflow": "hidden",
        "paddingBottom": 13,
        "paddingTop": 13,
      }
    }
    testID="RNE__SearchBar-wrapper"
  >
    <View
      style={
        Object {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
      testID="RNE__Input__view-wrapper"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#dcdce1",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "borderRadius": 9,
            "flexDirection": "row",
            "marginLeft": 8,
            "marginRight": 8,
            "minHeight": 36,
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              replaceTheme={[Function]}
              testID="RNE__ICON__CONTAINER_ACTION"
              updateTheme={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "#7d7d7d",
                        "fontSize": 20,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Ionicons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
        <TextInput
          allowFontScaling={true}
          editable={true}
          onBlur={[Function]}
          onCancel={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholderTextColor="#7d7d7d"
          platform="ios"
          rejectResponderTermination={true}
          style={
            Object {
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 6,
              "minHeight": 40,
              "overflow": "hidden",
            }
          }
          testID="RNE__SearchBar"
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={
        Object {
          "opacity": 0,
          "position": "absolute",
          "right": null,
        }
      }
      testID="RNE__SearchBar-cancelButtonContainer"
    >
      <View
        accessibilityRole="button"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={Object {}}
          testID="RNE__SearchBar-cancelButton"
        >
          <Text
            style={
              Object {
                "color": "#007aff",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              }
            }
          >
            Cancel
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`SearchBar wrapper component should match snapshot 1`] = `
<View
  testID="wrapper"
>
  <View
    style={
      Object {
        "backgroundColor": "#393e42",
        "borderBottomColor": "#000",
        "borderBottomWidth": 1,
        "borderTopColor": "#000",
        "borderTopWidth": 1,
        "padding": 8,
      }
    }
    testID="RNE__SearchBar-wrapper"
  >
    <View
      style={
        Object {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
      testID="RNE__Input__view-wrapper"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#303337",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "borderRadius": 3,
            "flexDirection": "row",
            "minHeight": 30,
            "overflow": "hidden",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              replaceTheme={[Function]}
              testID="RNE__ICON__CONTAINER_ACTION"
              updateTheme={[Function]}
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "#86939e",
                        "fontSize": 18,
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="RNE__ICON__Component"
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
        <TextInput
          allowFontScaling={true}
          editable={true}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholderTextColor="#86939e"
          platform="default"
          rejectResponderTermination={true}
          style={
            Object {
              "color": "#86939e",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 10,
              "minHeight": 40,
            }
          }
          testID="RNE__SearchBar"
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;
